src/test/resources/regressions/features/defunc/mutex-and-waitgroup1.gobra

```diff
@@ -0,0 +1,93 @@
+// Any copyright is dedicated to the Public Domain.
+// http://creativecommons.org/publicdomain/zero/1.0/
+
+package pkg
+
+import "sync"
+
+type Accum struct {
+	sum *int
+	m *sync.Mutex
+}
+
+pred AccInv(x *int) {
+	acc(x)
+}
+
+// TODO: debug this
+// requires acc(c) && acc(c.sum) && acc(c.m) && *(c.m) == sync.Mutex{}
+requires acc(c) && acc(c.sum) && acc(c.m) && c.m.state == 0 && c.m.stema == 0
+requires c.sum != nil
+ensures acc(c) && c.m.LockP() && c.m.LockInv() == (AccInv!<c.sum!>)
+func (c *Accum) Init() {
+	fold AccInv!<c.sum!>()
+	c.m.SetInv(AccInv!<c.sum!>)
+}
+
+// TODO: change names of vars and types
+// requires len(s) > 0
+/* Creates a worker thread per element in the slice and concurrently computes the sum of all the elements */
+requires s != nil
+func parallelSum(s []int) {
+	// { }
+	count@ := 0
+	m := &sync.Mutex{}
+	c := &Accum{sum: &count, m: m}
+	// { acc(c) && acc(c.sum) && acc(c.m) }
+	c.Init()
+	// { acc(c) && c.m.LockP() && c.m.LockInv() == AccInv!<c.sum!> }
+	w := &sync.WaitGroup{}
+	w.Init()
+	// {
+	//    acc(c) && c.m.LockP() && c.m.LockInv() == AccInv!<c.sum!> &&
+	//    w.WaitGroupP && !w.WaitMode()
+	// }
+
+	n := len(s)
+
+	invariant i >= 0
+	invariant i == 0 ==> w.WaitGroupP()
+	invariant i > 0 ==> acc(w.WaitGroupP(), 1/2) && acc(w.WaitGroupStarted(), 1/2)
+	// Martin suggested the following line instead of the previous one, with the goal of
+	// deleting the invariant `i >= 0`. However, that is insufficient to prove the loop invariant
```

Comment:
I came across the "stupid" invariants when working on examples as well. In my opinion the encoding of `for ... { ... }` should be made stronger such that invariants like `i >= 0` and `i <= n` are automatically inserted into the encoding

Reply:
Yes, that is definitely something that would be helpful!

Reply:
Felix has convinced me today that it's not as easy as I thought and would probably be quite heuristic based. E.g. inserting the mentioned invariants would only be correct if `i` is not changed in the loop body and/or captured in e.g. a closure.